File: src/main/java/com/oycm/algorithm/h/basic/Solution_10.java
package com.oycm.algorithm.h.basic;


public class Solution_10 {

    /**
     * 2917. <a href="https://leetcode.cn/problems/find-the-k-or-of-an-array/description/">找出数组中的 K-or 值</a> 1389
     *
     * @param nums
     * @param k
     * @return
     */
    public int findKOr(int[] nums, int k) {
        /*
        K-or 操作: 所有 nums[i] 某个比特位的存在至少 k 个 1，那么该 比特位的值为 1
        31 长的数组记录 nums[i] 对应比特位的数量

        cnt[i] >= k, 进行 或拼接
         */
        int[] cnt = new int[31];
        for (int num : nums) {
            int i = 0;
            while (num > 0) {
                if ((num & 1) == 1) {
                    cnt[i]++;
                }
                i++;
                num = num >> 1;
            }
        }
        int ans = 0;

        for (int i = 0; i < cnt.length; i++) {
            if (cnt[i] >= k) {
                ans |= 1 << i;
            }
        }

        return ans;
    }

    public int optimize(int[] nums, int k) {
        /*
        合并循环写法
         */
        int ans = 0;
        for (int i = 0; i < 31; i++) {
            int cnt = 0;
            for (int num : nums) {
                cnt += num >> i & 1;
            }
            if (cnt >= k) {
                ans |= 1 << i;
            }
        }

        return ans;
    }
}

File: src/main/java/com/oycm/algorithm/e/enum_right_log_left/basic/Solution_10.scala
package com.oycm.algorithm.e.enum_right_log_left.basic

object Solution_10 {

  /**
   * 2815. 数组中的最大数对和 1295
   * https://leetcode.cn/problems/max-pair-sum-in-an-array/description/
   *
   * 数 数位上的最大值：123 -> 3, 192 ->9
   *
   * @param nums 整数数组
   * @return 找出 和 最大 的一对，且这两个数 数位上最大的数字 相等
   */
  def maxSum(nums: Array[Int]): Int = {
    /*
    题解：取模找出数位的最大值
    */
    var ans = -1
    val cnt = Array.fill(10)(Int.MinValue)
    for (num <- nums) {
      var maxD = 0
      var x = num
      while (x > 0) {
        maxD = Math.max(maxD, x % 10)
        x /= 10
      }
      // 更新最大值 最小值和 max(nums(i)) 小于 -1
      ans = Math.max(ans, num + cnt(maxD))
      // 更新相同数位的最大值
      cnt(maxD) = Math.max(cnt(maxD), num)
    }

    ans
  }

  def main(args: Array[String]): Unit = {
    println(maxSum(Array(51, 71, 17, 24, 42)))
    println(maxSum(Array(1, 2, 3, 4)))
  }

}
